kong/conf_loader.lua

```diff
@@ -89,34 +89,76 @@ local HEADER_KEY_TO_NAME = {
 }
 
 
+local EMPTY = {}
+
+
+-- NOTE! Prefixes should always follow `nginx_[a-z]+_`.
 local DYNAMIC_KEY_NAMESPACES = {
   {
-    injected_conf_name = "nginx_http_status_directives",
-    prefix = "nginx_http_status_",
+    injected_conf_name = "nginx_main_directives",
+    prefix = "nginx_main_",
+    ignore = EMPTY,
   },
   {
-    injected_conf_name = "nginx_http_upstream_directives",
-    prefix = "nginx_http_upstream_",
+    injected_conf_name = "nginx_events_directives",
+    prefix = "nginx_events_",
+    ignore = EMPTY,
   },
   {
     injected_conf_name = "nginx_http_directives",
     prefix = "nginx_http_",
+    ignore = {
+      upstream_keepalive          = true,
+      upstream_keepalive_timeout  = true,
+      upstream_keepalive_requests = true,
+    },
+  },
+  {
+    injected_conf_name = "nginx_upstream_directives",
+    prefix = "nginx_upstream_",
+    ignore = EMPTY,
+  },
+  {
+    injected_conf_name = "nginx_proxy_directives",
+    prefix = "nginx_proxy_",
+    ignore = EMPTY,
+  },
+  {
+    injected_conf_name = "nginx_status_directives",
+    prefix = "nginx_status_",
+    ignore = EMPTY,
+  },
+  {
+    injected_conf_name = "nginx_admin_directives",
+    prefix = "nginx_admin_",
+    ignore = EMPTY,
   },
   {
     injected_conf_name = "nginx_stream_directives",
     prefix = "nginx_stream_",
+    ignore = EMPTY,
   },
   {
-    injected_conf_name = "nginx_proxy_directives",
-    prefix = "nginx_proxy_", -- TODO: nginx_http_proxy
+    injected_conf_name = "nginx_supstream_directives",
+    prefix = "nginx_supstream_",
```

Comment:
Should we call this `nginx_stream_upstream_directives ` instead. I was definitely confused at first what `nginx_supstream_directives` means and the `s` is easy to miss as well.

Reply:
@dndx we used to have `nginx_sproxy`, but the original plan was to follow nginx hierarchy:
- `nginx_*`
- `nginx_http_*`
- `nginx_http_proxy_*`

etc. and I implemented this first with that, but there was an issue that by the name of the injection we cannot determine which part of it is the context and which is the directive.

```
KONG_NGINX_HTTP_PROXY_SSL_CIPHERS=y -> proxy_ssl_ciphers OR ssl_ciphers?
```

How can you know what user means about it? I originally even went with longest match semantics, and whitelists and blacklists etc. But the problems just popped up.

Thus we changed all injections in form:
```
KONG_NGINX_<SINGLE-WORD>
```

This kept things also quite nice on backward compatibility as we already had `NGINX_HTTP_*`, `NGINX_PROXY_*`, `NGINX_SPROXY_*`. And there the `S` prefix was taken to others too, e.g. `NGINX_SUPSTREAM_*´. Yes, I know it is not as nice as that hierarchical, but at least it is shorter.

Reply:
Okay, I see that, thanks for the detailed explanation!